scipy.signal: Add the class dlti for linear discrete-time systems. Closes #2912

befelix opened the pull request:
This extends the current system representation to discrete-time systems. The overall idea is to have a master class ```LinearTimeInvariantSystem```, that combines all the functionality, but is not actually designed to be called by the user. Below that are ```lti``` and ```ltid```, which can be called and return the corresponding subclass (e.g. ```StateSpace```). What's new now, is that these subclasses don't inherit from ```lti``` anymore. Instead they return an instance of ```StateSpaceContinuous``` or ```StateSpaceDiscrete```, depending on the input arguments. These last subclasses inherit from ```StateSpace``` and ```lti``` or ```ltid```, which enables the ```StateSpace``` class to decide at runtime from which of the two it should inherit. While these last names seem cumbersome, you never actually have to type them, use them, or even see them (unless you call print on a system).

In summary: Everything stays the same for continuous-time systems. But for discrete-time systems there is now ```ltid``` and you can call ```StateSpace(1,1,1,1, 0.05)```, where 0.05 is the sampling time.

One drawback: 
- Since lti/ltid classes share functionality, some functions had to be moved to the same file in order to avoid circular imports. Specifically, the files dltisys.py and cont2discrete.py were removed.

This ended up being a rather large commit and took time to get to a state that I think is good. It would be great if you could invest some time to review it. Imho this is the last big user-facing change to the api that needs to happen in order to make linear systems in scipy useful.

befelix commented:
I don't know why the test failed, especially since that fail occurs in a seemingly completely unrelated file and the test passes when calling the test-file directly. I'll have a look, but just want to point out that the code actually works :)

FRidh commented:
That's an impressive PR. I'll have a closer look at it again later, but for now I have one question. Why is there
```
@property
    def system_class(self):
```
If someone needs the actual class they might as well use obj.__class__

befelix commented:
Oh, I think that's just some leftover code from a previous design I had for this. Ignore it for now, I'll remove it in time.

FRidh commented:
This PR has been around now for quite some time without much feedback.

In #2912 there was a discussion on `fs` vs `sampling_time`. Most people seemed to prefer `sampling_time` which is implemented here. I'm fine with that.

I think it would be good if this would be merged as soon as possible since then we can add `lfilter`/`filtfilt` variants to it as well as `SecondOrderSections(LinearTimeInvariant)`.

befelix commented:
Yes, part of this is my fault since I was extremely busy ever since I send in the PR and will be for another week, but afterwards I'll take this up more actively again. As for the sampling_time / fs / dt argument I don't have a strong preference. If we stick to what is already used I would probably prefer dt over fs. 

I will sort out the merge conflicts as well as your comment about the system_class as soon as I have time again (after the 16th). Though it would be great if some people could review the code properly in the meantime.

e-q commented:
I definitely think this is a good feature. I'll take a closer look into the code when I have some time. Why is the Bode method commented out?

As for the parameter choice: this feels a bit bikeshed-y. I prefer specifying the sampling frequency, but its more important to just be consistent. MATLAB isn't a good model for this; e.g. creating an LTI system takes the timestep, whereas some commands like `bilinear` take the sampling frequency.

befelix commented:
Alright, rebased the code, removed the system_class method, uncommented the bode plot code. Still not sure what to use instead of sampling time.
@e-q The argument against the sampling frequency was that it is infinity for continuous time systems, which is somehow not that nice. (This was brought up in #2912)

e-q commented:
Hey @befelix, it's been entirely too long, but I started to try and review this but got caught up in rebase hell...

I'm willing to work through it, as I really want to see this functionality in scipy, but it may be a little slow going, since I'll be learning what you did as I rebase. 

If you have the time and willpower, though, it could be nice if you rebase it, and we continue on this PR (maybe my git-fu isn't strong enough, but I think that if I do the rebasing, we'd have to open a new one..)

Let me know what you think!

befelix commented:
Thanks for agreeing to review this. I'll try to clean it up and rebase it
within the next week.

On Thu, 3 Mar 2016 01:47 Eric Quintero, <notifications@github.com> wrote:

> Hey @befelix <https://github.com/befelix>, it's been entirely too long,
> but I started to try and review this but got caught up in rebase hell...
>
> I'm willing to work through it, as I really want to see this functionality
> in scipy, but it may be a little slow going, since I'll be learning what
> you did as I rebase.
>
> If you have the time and willpower, though, it could be nice if you rebase
> it, and we continue on this PR (maybe my git-fu isn't strong enough, but I
> think that if I do the rebasing, we'd have to open a new one..)
>
> Let me know what you think!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy/scipy/pull/4881#issuecomment-191513081>.
>

befelix commented:
@e-q I've rebased the code. Let me know if you find anything strange.

e-q commented:
Awesome, I won't be able to look through until next week, but I look forward to it!

e-q commented:
I'm still chewing on this, but here are a few things that have occurred to me so far:

I feel like we don't have to have separate functions for `freqresp` and `dfreqresp`, if the function can just figure out whether its discrete or continuous, and call `freqs` or `freqz` appropriately. This reduces duplicated code, and fixes the `ltid.bode` method (which currently fails). In any case, even if this is kept separate, there are currently no tests for `dfreqresp`.

I think all of the `LinearTimeInvariant` methods like `num`, `den` and so on should use the `as_tf` kind of functions instead of `to_tf`. We've seen this cause unneeded latency in places like gh-5613.

Inspecting an `lti` or `ltid` object in IPython via `sys?` is less helpful with the new "Dummy subclass" docstring. I would prefer duplicating the long docstring from the parent class, with appropriate modifications.

e-q commented:
Also, `lutists.py` is getting huge, currently 3369 lines. Maybe the conversion and simulation functions can be broken out into files like `lti_conversion.py` and `lti_sim.py`?

e-q commented:
Looking at [the codecov report](https://codecov.io/github/scipy/scipy/scipy/signal/ltisys.py?ref=a217347cf779a266c94a4c1df055f5b19b6df724) I don't see any other big gaps in tests right now.

e-q commented on scipy/signal/ltisys.py:
```diff
@@ -2506,3 +2891,479 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+        is_ss_input = True
+    else:
+        is_ss_input = False
+        system = system.to_ss()
```
`as_ss`

befelix commented:
Thanks a lot for the feedback!

The seperation between `freqresp` and `dfreqresp` is necessary for consistency. All the `lsim` related functions have the split and the discrete time versions take the sampling time as part of the system-tuple. Unfortunately this makes it difficult to differentiate if the system is a discrete-time transfer function with sampling time or a zeros-poles-gain system in continuous time.

I'll add the tests, but it will probably be some time until I can. It could be an option to split this feature from this PR in order to make it less monolithic.

The reason why the `_as_ss' is not necessary for the LinearTimeInvariant system properties is that, for example, the `StateSpace` class will overwrite the respective `A`, `B`, `C`, and `D` properties with the actual system matrices (no conversion).
In any case, my personal long-term goal would be to see these cross-class properties depricated. They are really inefficient to compute, encourage bad behavior, and can now be easily accesses by something like `system.to_ss().A`, which makes it clear to the user that a system conversion is taking place. 

I agree that it would probably be nice to have a separate file only for the system representations and to have conversions and simulations somewhere else. Though, since this is more of an esthetic question, for now I think that splitting up the file would only make the diff-log of this PR even more difficult to navigate. 

Let me know if you find anything else!

e-q commented:
Thanks for the clarifications.

> The seperation between freqresp and dfreqresp is necessary for consistency. All the lsim related functions have the split and the discrete time versions take the sampling time as part of the system-tuple. 

I see what you're saying. Really, this is one of the reasons why I'm not a huge fan of the system-as-tuple / argument counting API. 

It makes sense to me that the discrete and continuous lsim functions are separate, since the nature of the computation is different. I personally don't feel like this doesn't have to extend to lower level things like `freqresp` and `bode`, though I understand that this is impractical with the current API design.

What would be your take on an API where there is no `ltid` class, but instead `lti` had a keyword argument `Ts` or `dt` that defaults to 0? Of course, one problem with this is that `dlsim`, etc., already use the (num, den, dt) system-tuple arguments, but is there another strong reason why we shouldn't try something like this?

e-q commented:
Also, for the current API, we're missing a `dbode` method. `ltid.bode` currently calls `bode` and thereby the continuous-time `freqresp`.

befelix commented:
> Also, for the current API, we're missing a dbode method. ltid.bode currently calls bode and thereby the continuous-time freqresp.

Oh true, I guess this goes back to the question what the proper way is here. `boded` is sound strange, but the alternative is either to break backwards-compatibility (takes a lot of time and beyond the scope of this PR), or to have `bode` work for continuous-time tuples, continuous-time systems, and discrete-time systems. Though that might be confusing as well. 

I think my current preference would be to remove `freqrespd` and `bode` from the `dlti` class and this PR and to move this discussion into a seperate thread. Should make an eventual merge easier to achieve. What do you think?

> What would be your take on an API where there is no ltid class, but instead lti had a keyword argument Ts or dt that defaults to 0? Of course, one problem with this is that dlsim, etc., already use the (num, den, dt) system-tuple arguments, but is there another strong reason why we shouldn't try something like this?

I actually had this as part of the `LinearTimeInvariant` class at some point, but didn't like it. It seemed much saner to me to use the correct subclass that fits the model directly, rather than relying  on the voodoo-magic of `__new__` and having yet another entry point for system creation. 

There are discrete-time specific attributes (e.g., at some point there should be an `sos` system class). It is of course possible to have `lti` take the place of what `LinearTimeInvariant` currently serves, but I don't see a way around the current structure with separate subclasses for continuous and discrete time.

Having a consistent API is, in my opinion, the most important feature of any library. So while a long-term goal could be to have a discussion about changing the API, I think this is beyond the scope of this PR.

That being said, I'm of course open to concrete proposals of how to do this better and I really appreciate you taking the time to review this!

e-q commented:
I think that `dfreqresp` and `dbode` should land in this PR, it's awkward to have the discrete classes without them, in my opinion. In terms of backwards-compatibility and easy inclusion in v0.18, your current approach is probably indeed probably the best. 

I suppose my underlying concern in my replies above was to avoid doing something here the makes it harder for scipy.signal to eventually land at a cohesive API, since there are thoughts floating around about a general overhaul of the entire LTI filtering API (as is partially discussed in gh-3259 and probably other places). This is definitely outside the scope of this PR, but the way we handle discrete vs. continuous should definitely be a factor in whatever ends up happening. 

I'm going to throw out some pings to people who may have more well-founded opinions / better ideas than I. @rgommers @endolith @WarrenWeckesser @ewmoore  @Eric89GXL

Eric89GXL commented:
> I suppose my underlying concern in my replies above was to avoid doing something here that makes it harder for scipy.signal to eventually land at a cohesive API

I haven't worked with these classes, so take my comments with a grain of salt.

My sense is that, given the tension between wanting a cohesive API (and not hindering progress toward it) and wanting to add features consistent with the current API, we'll have to side with adding features. It will make the eventual merge more difficult, but it would be a shame to miss out on good concrete improvements in the meantime that people will actually benefit from. If we were on the verge of having our API sorted out and unified -- say, within a couple of months, with a PR in progress with some momentum -- I'd say we should hold off. But I'm pretty sure we're not there :) In the meantime my vote is to try to fit with the current API first, and try to reasonably minimize future challenges for unification in the process.

Since I haven't used these specific classes, I don't have any useful comments on the other issues (`dbode`, `boded`, etc.).

e-q commented:
Hey @befelix, if you don't have time to move this forward, I can take this over and squash and commit this along with `dbode` and some more tests.

befelix commented:
@e-q, that would be really helpful. It will probably be another week before I will have time to add the finishing touches. If you do have time before that it would be great if you could keep our respective contributions separate of course. Thanks!

befelix commented:
Okay, I have a question about the discrete-time transfer function. In the code, I have assumed that discrete-time transfer functions numerators&denominators are given the same way as in Matlab and for scipy's continuous-time systems (`1 + 2 z^1 + 3 z^2 = [1, 2, 3])`. This should allow us to reuse all the conversion functions (eg., `tf2ss`). Unfortunately, I just saw that `freqz` and the discrete-time scipy filters take the system in a different form: (`1+ 2 z^(-1) + 3 z^(-2) = [1, 2, 3]`). It's not difficult to convert between the two representations.

The question now is: which representation should we choose? One is consistent with the conversion functions, the other one with the discrete-time filters.

endolith commented:
@e-q:

> avoid doing something here that makes it harder for scipy.signal to eventually land at a cohesive API, since there are thoughts floating around about a general overhaul of the entire LTI filtering API (as is partially discussed in gh-3259 and probably other places).

Isn't this part of that?  gh-3259 is just about wanting functions to take a system as input instead of separate lists of coefficients, and this provides a system object for that purpose.

> I prefer specifying the sampling frequency

Yeah, me too, since it would almost always be written `dt=1/fs` anyway, and `fs` is usually an exact integer.  Does it have any practical effects?  Like sample rate conversion of integers works out exactly 48000 * 147//160 = 44100 , while working with dt gives 44100.00000000001 Hz.

@befelix 

> The argument against the sampling frequency was that it is infinity for continuous time systems, which is somehow not that nice.

But continuous time system objects don't store the sampling frequency at all?

@e-q 

> MATLAB isn't a good model for this; e.g. creating an LTI system takes the timestep, whereas some commands like bilinear take the sampling frequency. 

well scipy's already inconsistent, too

- scipy [signal.bilinear](http://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html) uses fs 
- scipy [signal.cont2discrete](http://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.cont2discrete.html) uses dt

> I feel like we don't have to have separate functions for freqresp and dfreqresp, if the function can just figure out whether its discrete or continuous

l like the sound of that at first, but then the `w` argument will mean different things in each case.

Similarly `residue` and `residuez` could be combined if they only accepted system objects and could tell the difference between analog and digital, but they should continue to accept list of coefficients, too, in which case they should stay separate.  Also there's already `impulse` vs `dimpulse`, etc.  So I'm fine with leaving digital and analog functions separate.

> and call freqs or freqz appropriately.

(Well in the future it should not call `freqs` or `freqz` at all since they use `tf` representation which is numerically lossy. https://github.com/scipy/scipy/issues/5648)

befelix commented:
@endolith 
> Yeah, me too, since it would almost always be written dt=1/fs anyway, and fs is usually an exact integer. Does it have any practical effects? Like sample rate conversion of integers works out exactly 48000 * 147//160 = 44100 , while working with dt gives 44100.00000000001 Hz.

Well this is where different communities have different notions. Signal processing will typically use fs, while control people will use dt, since for them discrete-time systems are often just continuous-time systems that are analytically integrated assuming a constant input. 

My goal was to avoid to avoid this whole discussion - the API in the PR is designed to be consistent with all the functions that already existed for discrete-time systems. Hence, I used `dt` and the discrete-time TransferFunction takes numerators and denominators in the form `z^2 + 3z + 5 = [1, 3, 5]`. 

My personal preference for a good api would be to have a keyword argument for sampling time or frequency (or even both). That way one could do things like
`TransferFunction(1, [2, 3])`
`TransferFunction(1, [2, 3], sampling_time=0.1)`
`TransferFunction(1, [2, 3]), frequency=10)`

I guess that would still be possible given that this is new code, but it would just add to the inconsistency at this point. 

@e-q Found some time today to add the tests after all. Thanks for your offer to help though!
<br>
**Open questions:**
* `TransferFunction` in terms of `z` or `z^-1` (`dlsim` and the likes assume `z`).
* Verbosity of parameters: `sampling_time` vs `dt`
* API consistency (at least for discrete-time system functions) vs 'improving' stuff
* Should dbode return frequencies in rad/s?

e-q commented:
I've come around to appreciating separate classes/functions for discrete and continuous, especially given that only the discrete systems can be used to filter data. (Maybe it's silly to care, but maybe the class should be named `dlti`, since all of the other discrete-time functions use `d` as a prefix rather than as a suffix.)

I do like the idea of being able to use both `dt` and `fs`, since we likely have users that are used to each. Naturally, this sets the expectation of enabling a similar capability and implies a fair amount of work, but maybe is warranted given the amount of things that need to be done to get to a v1.0 level of consistency anyways. This could probably be done in a quick follow-up PR though.

@befelix
- I'm used to seeing inverse z powers. I think our conversion functions need some work anyways, so I wouldn't be inclined to give their precedents too much weight.  
- I think `dt` as a name is fine, we already use `fs` for sampling frequency. 
-  Yeah, I think rad/s is appropriate, which means `dt` is needed. This actually strikes me as ok, so that the input system tuple to `dbode` is the same as one would use for `ltid`. (The standard WW comment: really it should be rad/[time unit] since `dt`/`fs` doesn't have to be in seconds/Hz to be valid). I suppose dbode and its kind could return rad/sample without knowing the sampling time, but the same is achieved with `dt=fs=1`.

@endolith
- I agree on the the `freq(s/z)` point, see gh-6059

Zooming out a bit, I've been thinking that we need to get consensus on a concrete vision of what we want the `scipy.signal` API/landscape to look like, so that we can attack it in a more systematic manner, rather than trying to come up with the right design decisions on a PR-by-PR basis. Where would be the right place to do this? @endolith started a to-do list on the github wiki... maybe I'll send an email out to scipy-dev to collect more ideas/opinions, which can then be whittled down to something tractable.

Eric89GXL commented:
> Where would be the right place to do this? @endolith started a to-do list on the github wiki... maybe I'll send an email out to scipy-dev to collect more ideas/opinions, which can then be whittled down to something tractable.

I think the best approach would be for someone to take the time to:

1. Look at this issue and other related ones (second-order sections and related ones come to mind) where we've talked about this
2. Synthesize one or a couple different API proposals based on what's been discussed already, along with pros/cons that have been discussed
3. Make a GitHub issue dedicated to it (better than Wiki because people can comment)
4. Email the listserv to solicit opinions

A big problem with discussing these issues is that the information is scattered, and there are a lot of potential factors involved. So if someone takes the time to put together a proposal or a couple of proposals it could go a long way I think. I don't think it makes much sense to try to collect opinions from people until these ideas are distilled to such clear, succinct presentations.

befelix commented:
@e-q @Eric89GXL 
>  I've been thinking that we need to get consensus on a concrete vision of what we want the scipy.signal API/landscape to look like, so that we can attack it in a more systematic manner, rather than trying to come up with the right design decisions on a PR-by-PR basis.

I agree that would be helpful. Unfortunately I don't use the filtering part of the library enough to be helpful on this, but I'd be happy to be part of the discussion. Here is my opinion on this. I think it should work for most usecases, but like I said: I'm not familiar with all the functionality.

General
* There should be not tuple representations of systems. Functions that take arbitrary system tuples followed by a conversion to `lti`/`ltid` should take only `lti`/`ltid` systems instead. Potentially, they should also only take systems for which there is code around, rather than implicitly doing conversions inside. That is, if the code only works for `StateSpace` systems, then only `StateSpace` systems should be accepted as inputs. At the very least, there should be a warning about the conversions in the docstring.
* All transfer functions should be in terms of consistent variables, probably `z**-1`. Matlab's solution here is to take polynomials in `z` as inputs, but `num`/`den` are returned in terms of `z**-1`. Let's be better than that.

LinearTimeInvariant: 
* `lti` / `ltid` should not be callable. Instead, the subclasses should be used directly. This could be important if the subclasses become more sophisticated and take optional keywords.
* (Optional) The system subclasses should not accept instances of other classes as input. Instead, the `to_ss()`/`to_tf()`/`to_zpk()` methods should be used. As a result, the API would read something like `StateSpace(a, b, c, d, dt=None)` rather than `StateSpace(*system, **kwargs)`, which seems nicer (personal preference).

e-q commented:
I like how this is shaping up @befelix! I'm on board with moving away from system-tuples altogether, though I will take @Eric89GXL's suggestion and start piecing things together for a discussion of API proposals.

I'm going to throw a question out here: Is it desirable to allow the creation of discrete-time system without defining the time step? I haven't personally run into this kind of use case, but that doesn't mean all that much...

Since `dfreqresp` is returning rad/sample, it doesn't need to know `dt` at all, which made me think of this.

Eric89GXL commented:
At least in many filtering applications you don't need to know the time step. So I'd say do without it where we can, or default to `1.` which will just leave units in rad/samp anyway. Or maybe `dfreqresp` should take the sample rate into account somehow... not sure.

endolith commented:
>  keyword argument for sampling time or frequency (or even both). 

Well it doesn't really matter if the *argument* is `dt` or `fs`; it's trivial to do `fs=1/dt` or `dt=1/fs`, so I don't think it needs arguments for both.  I'm just concerned about it using accurate values internally.  Usually sample rates are whole numbers and dt are not, so when combining systems that use dt, you'll be testing floats for equality, for instance?

> TransferFunction in terms of z or z^-1 (dlsim and the likes assume z).

I would say both, since they're both used in different communities.  I started to add this to tf2zpk, for instance.  https://github.com/scipy/scipy/pull/4872

> Verbosity of parameters: sampling_time vs dt 

`dt` or `fs` is better than `sampling_rate` or `sampling_time` in my opinion.  docstring will explain it

Also how would these filter objects be designed?  Like how does `butter(N, Wn, analog=False, output='dlti')` know what sample rate to use in the object?

I've been thinking the different normalized frequencies used in different functions (radians / sample in some functions vs half-cycles / sample in others) could be harmonized by adding an fs= parameter anyway:

    fs = 48000
    f = linspace(1000, 3000, 100)

    # Currently different
    b, a = butter(4, 2000/(fs/2))
    w, h = freqz(b, a, 2*pi*f/fs)

    # Simpler to work with
    b, a = butter(4, 2000, fs=fs)
    w, h = freqz(b, a, f, fs=fs)

I've said somewhere before that I think the filter objects should be callable so it's possible to just create a filter and filter a signal with it, without caring how it's implemented.  It automatically uses the most accurate methods internally (generating as ZPK and then filtering using SOS, for instance).  Something like this:

    sig = randn(48000)
    # 2 kHz lowpass filter object
    lowpass = butter(4, 2000, fs=48000, analog=False, output='lti')
    filtered_sig = lowpass(sig)

> There should be not tuple representations of systems.

What's the reason for removing them?  Functions like `freqresp` would not longer accept tuples?

> At least in many filtering applications you don't need to know the time step. 

I didn't realize [python-control already has StateSpace and TransferFunction classes](http://python-control.readthedocs.org/en/latest/classes.html).  Are they different enough that we can't use them?
    
In theirs
 - dt=float is a discrete-time system
 - dt=True is a normalized-frequency discrete-time system(?)
 - dt=None is a continuous-time system

befelix commented:
> 
>     keyword argument for sampling time or frequency (or even both).
> 
> Well it doesn't really matter if the argument is dt or fs; it's trivial to do fs=1/dt or dt=1/fs, so I don't think it needs arguments for both. I'm just concerned about it using accurate values internally. Usually sample rates are whole numbers and dt are not, so when combining systems that use dt, you'll be testing floats for equality, for instance?

I actually had an implementation of having both `dt` and `fs` keywords, but then removed it since it seemed somewhat messy. Maybe a more appropriate way would be to have another keyword that just gives the unit of the sampling time - so people could pass Hz, s, etc.


> 
>     TransferFunction in terms of z or z^-1 (dlsim and the likes assume z).
> 
> I would say both, since they're both used in different communities. I started to add this to tf2zpk, for instance. #4872 

I agree that it should be possible to specify both, but the default value should be consistent, and the `num` `den` properties of the transfer function should correspond to that default. Otherwise the API becomes super confusing if for every function you first have to check which convention it expects. Btw, this is already a problem since dltisys expects `z`, but the filtering usually `z**-1`.

>     There should be not tuple representations of systems.
> 
> What's the reason for removing them? Functions like freqresp would not longer accept tuples?

My motiviation for removing them is that they create a lot of duplicate code. For example, if you now say that we could specify transfer functions both in terms of `z` and `z**-1` then basically you need to add the corresponding keywords to any function that takes a linear system tuple as input. That's why classes exist in the first place - so that all the information about the system is in one place. 

And yes, `freqresp` would not accept tuples anymore either if it was only up to me :)

> I didn't realize python-control already has StateSpace and TransferFunction classes. Are they different enough that we can't use them?

So they follow the route of having one class that is the same for continuous or discrete time systems. There are a couple of reasons why I don't like this
* Methods that only work for continuous-time systems will have to raise an error if called from a discrete-time system (e.g., `to_discrete()`). With the different classes in this patch this is not an issue.
* If statements in every function to check if we're dealing with continuous or discrete time systems seem less clean to me.

Beyond that they use the system tuples and have a very matlab-focused style. Though it's good to know that they have operators (+/-/*) for all systems already implemented. I have a follow-up PR to this one that implements this for StateSpace systems, but since the code for all systems is already in this library one can probably draw inspiration from there.

> 
> In theirs
> 
>     dt=float is a discrete-time system
>     dt=True is a normalized-frequency discrete-time system(?)
>     dt=None is a continuous-time system

I like the `dt=True` option, especially since it evaluates to 1 and all the code should work for this without modification.

e-q commented:
Ok, I've started an issue about the broader issues at play. Let's focus on what we need to bring this PR through now. I like the `dt=True` suggestion, there should be something in the docstrings to talk about that. 

Otherwise, I don't think there is anything else blocking this on my end. I'll take another look through it, but is there anything else you want to see done/changed here @befelix / @endolith / @Eric89GXL?

Eric89GXL commented:
Nothing from my end.

e-q commented:
@befelix: Do you have any time to make the adjustments for the `dt=True` option? Once that's done, I think this can be merged.

befelix commented:
@e-q  Sorry for being so quiet. For me the main blocking issue is the question of whether the default should be ```z``` or ```z**-1``` for the TransferFunction. Once merged this is difficult to change and the default should ideally be consistent between functions in the long term.

I'll take care of ```dt=True``` soon, though my guess would be that only docstring changes are required, since ```True``` is pretty much the same as 1.

befelix commented:
Alright, `dt=True` is now part of the docstrings. Didn't require any changes to the codebase.

Is there any hope of getting this into 0.18? What more is needed to get this in?

Eric89GXL commented:
I'll mark it optimistically for 0.18, since it seemed like the `dt` was the only lingering issue. @e-q will you have time to look again and merge if you're happy?

endolith commented on scipy/signal/ltisys.py:
```diff
@@ -555,18 +336,163 @@ def bode(self, w=None, n=100):
+        The following gives the number of arguments and the corresponding
+        subclass that is created:
+
+            * 3: `TransferFunction`:  (numerator, denominator)
```
shouldn't these all refer to TransferFunctionDiscrete, etc?

endolith commented on scipy/signal/ltisys.py:
```diff
@@ -555,18 +336,163 @@ def bode(self, w=None, n=100):
+    both the numerator and denominator should be specified in descending
+    exponent order (e.g., ``z^2 + 3s + 5`` would be represented as ``[1, 3,
+    5]``).
+
```
can you add Examples?  my first attempt was 

```
ltid([1,2], [4, 5], 1)
Out[23]: 
ZerosPolesGainDiscrete(
array([1, 2]),
array([4, 5]),
1,
dt: True
)
```
which was not what I expected.  my second attempt was 

```
ltid(([1,2], [4, 5]), 1)
Out[24]: 
TransferFunctionDiscrete(
array([[ 1.,  2.],
       [ 4.,  5.]]),
array([ 1.]),
dt: True
)
```
which is also not as expected

befelix replied:
So the first one is a result of taking `dt` as a keyword argument for the system-subclass (e.g., `StateSpace`). So it would have been `ltid([1,2], [4,5], dt=1)`. I guess this doesn't make much sense for `ltid`, so I'll change that do be `ltid([1,2], [4,5], 1`.

The second is mostly a result of old code that behaved this way. Could change it for the discrete-time system, but then things would  behave differently in continuous and discrete-time systems. MIMO systems are not really supported at the moment.

befelix replied:
Ah, I remembered wrongly. The reason for the `dt` keyword argument was that if it's not specified we default to a discrete-time system with `dt=True` otherwise. So it makes sense to keep it as it is currently.

A future PR of mine will try to deprecate being able to call `lti` and `ltid` directly anyways in favor of using the subclasses directly. The current way has too many access paths.

endolith replied:
You should clarify that it must say `dt=` explicitly or else it will be considered part of the system.

endolith commented:
```
TransferFunction([1, 2], [4, 5], dt=1)
Out[28]: 
TransferFunctionDiscrete(
array([ 0.25,  0.5 ]),
array([ 1.  ,  1.25]),
dt: 1
)
```

Does it have to normalize the coefficients?  I would expect it to keep them in the format that was entered, I think.

endolith commented on scipy/signal/ltisys.py:
```diff
@@ -889,30 +1051,134 @@ def to_ss(self):
+    Changing the value of properties that are not part of the
+    `ZerosPolesGain` system representation (such as the `A`, `B`, `C`, `D`
+    state-space matrices) is very inefficient and may lead to numerical
+    inaccuracies.
```
Can you clarify what to do instead if you want to modify them?  Like convert to a StateSpace system first?

befelix commented:
> Does it have to normalize the coefficients? I would expect it to keep them in the format that was entered, I think.
So the old `lti` class normalized the coefficients. I think it's anooying too, but nothing to be done about this without breaking backwards compatibility. I think we could add a `normalize=True` keyword argument to the `TransferFunction` class, but this PR really only changes the way LTI systems are structured, not the way the classes work.

> Was it going to have a system.fs property to get the sampling rate as well?
Left that for a future PR since this is already monolithic and it's not clear whether there should be a `fs` keyword or something similar to the `unit` keyword of Matlab.

e-q commented:
- I agree we can get this into 0.18, and I really want to make it happen
- @endolith's perspective is a good barometer of what our experienced users will face. All of the classes' docstrings should have a concise instantiation example. 
- Options for not normalizing, adding a property for `fs`, etc. are not blocking for this PR. Some small follow-ons would be fine as far as I am concerned. 

Thanks for sticking it out @befelix!

e-q commented:
Also, for the record, my vote is for negative power polynomials - i.e. `z**-1` form - since this is something I see more often in the literature.

befelix commented:
Alright, I think most of @endolith's concerns should be adressed now. Also renamed `ltid` to `dlti`, which was mentioned way back but somehow forgotten.

Only question I have is whether we should expose the specific subclasses, such as `StateSpaceContinuous`, to the user. In the end there is no benefit to calling them over `StateSpace`, so it might just make the API seem a bit cluttered.

> Also, for the record, my vote is for negative power polynomials - i.e. z**-1 form - since this is something I see more often in the literature.

@e-q I tend to agree. Only downside is that it doesn't play nice with the discrete-time simulation functions anymore (they take `z`). I'll see if I can make a commit for this tonight.

befelix commented:
@e-q There's an implementation of `z**-1` for Polynomials at https://github.com/befelix/scipy/commits/discrete_lti_zinv (latest commit). The only thing missing is a probler implementation of `normalize` for discrete time systems, but the hack around that is not too bad.

The main reason why I haven't merged this with the PR yet is that I'm a bit concerned how not-nicely this now plays with all the other LTI functions
* Conversion functions only work for `z`, so there's a bit of back-and-forth going on there
* The simulation functions take `num`/`den` with `z`, so transfer functions are now handled as a special case

A consequence of this is that all these functions behave differently when called indirectly from the class or directly. For example, `TransferFunction(num, den).to_ss()` and `StateSpace(*tf2ss(num, den))` yield different results.

I think overall it may be better to stick with `z` instead of `z**-1` for now, and migrate all functions together at some point in the future. What do you think?

e-q commented:
Hm. Given the extent to which it implies modifying even more code, I'm inclined to agree that `z` is the better choice for now. We should open up a issue about this though, so that we don't lose track of it.

befelix commented:
> Hm. Given the extent to which it implies modifying even more code, I'm inclined to agree that z is the better choice for now. We should open up a issue about this though, so that we don't lose track of it.

Alright, I'll open an issue together with some follow-up PRs once this is merged.

So I think I adressed all of @endolith's concerns except for the question whether to expose subclasses such as 'StateSpaceDiscrete' to the user.

Let me know if there are any other concerns/questions. I would really like to wrap this up for 0.18.

e-q commented:
I think there enough classes exposed without throwing in the whole zoo. By my count, we'll expose:

- `lti`-> Any of the continuous system representations
- `dlti` -> Any of the discrete system representations
- `TransferFunction` -> Either continuous or discrete tf systems
- `ZerosPolesGain` -> Either continuous or discrete zpk systems
- `StateSpace` -> Either continuous or discrete SS systems

e-q commented:
I'm happy with this as it stands, but I'll give people a couple of days to comment further. Barring objections, I'll merge it this weekend.

rgommers commented:
> Only question I have is whether we should expose the specific subclasses, such as StateSpaceContinuous, to the user. In the end there is no benefit to calling them over StateSpace, so it might just make the API seem a bit cluttered.

I agree, better not to expose more classes.

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -678,119 +704,355 @@ def num(self):
+    --------
+    >>> from scipy import signal
+
+    # Transfer function: H(s) = 5(s - 1)(s - 2) / (s - 3)(s - 4)
```
This doesn't render correctly. There should always be a blank line between lines terminal in/output and text. And then the ``#`` isn't needed. See https://circle-artifacts.com/gh/scipy/scipy/99/artifacts/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.ZerosPolesGain.html#scipy.signal.ZerosPolesGain for rendered html on this PR.

This comment applies to many docstrings in this PR.

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -678,119 +704,355 @@ def num(self):
+            * 1: `lti` or `dlti` system: (`StateSpace`, `TransferFunction` or
               `ZerosPolesGain`)
             * 3: array_like: (zeros, poles, gain)
+    dt: float, optional keyword argument
```
``optional keyword argument`` --> ``optional``

rgommers replied:
applies to a number of docstrings

rgommers commented on scipy/signal/tests/test_dltisys.py:
```diff
@@ -264,5 +280,370 @@ def test_more_step_and_impulse(self):
 
 
+class TestLtid(object):
+    def test_ltid_instantiation(self):
```
In the two lines above, change ``ltid`` to ``dlti``

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -527,6 +191,141 @@ def D(self, D):
+    system representation (such as the `zeros` of a `StateSpace` system) is
+    very inefficient and may lead to numerical inaccuracies. It is better to
+    convert to the specific system representation first. For example, call
+    `sys = sys.to_zpk()` before accessing/changing the zeros, poles or gain.
```
double backticks for code

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -588,41 +382,245 @@ def freqresp(self, w=None, n=10000):
+    Parameters
+    ----------
+    *system: arguments
+        The `dlti` class can be instantiated with either 3, 4 or 5 arguments.
```
There's some asymmetry with ``lti`` here - ``lti`` docstring says "2, 3 or 4 arguments". Implementation and length checks in ``__new__`` look the same

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -588,41 +382,245 @@ def freqresp(self, w=None, n=10000):
+    system representation (such as the `zeros` of a `StateSpace` system) is
+    very inefficient and may lead to numerical inaccuracies.  It is better to
+    convert to the specific system representation first. For example, call
+    `sys = sys.to_zpk()` before accessing/changing the zeros, poles or gain.
```
double backticks

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -588,41 +382,245 @@ def freqresp(self, w=None, n=10000):
+        Calculate Bode magnitude and phase data of a discrete-time system.
+
+        Returns a 3-tuple containing arrays of frequencies [rad/s], magnitude
+        [dB] and phase [deg]. See `scipy.signal.dbode` for details.
```
this can simply be `` `dbode` ``

rgommers replied:
you only have to add namespaces if you're linking from a function *not* in the same namespace.

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -1024,37 +1490,168 @@ def to_tf(self, **kwargs):
+            The current system (copy)
+
+        """
+        return copy.deepcopy(self)
```
Why does this need to be a copy? I'd expect ``return self``

befelix replied:
Because the `lti` class behaves the same way. I think back when we added this the motivation was for `StateSpace(sys)` to work the same way as `sys.to_ss()`. In hindsight I probably would prefer not copying as well.

Notice that there are the `_as_ss()` functions as part of the class to avoid unecessary copying. These could be made part of the public API as well.

rgommers replied:
No need for a second method I think, that makes the API a bit ugly. 

> Because the lti class behaves the same way. 

OK, I see where this came from now. It's not a major problem, so let's just live with it.

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -588,41 +382,245 @@ def freqresp(self, w=None, n=10000):
+                return StateSpaceDiscrete.__new__(StateSpaceDiscrete, *system,
+                                                  **kwargs)
+            else:
+                raise ValueError('Needs 2, 3 or 4 arguments.')
```
``3, 4 or 5`` not ``2, 3 or 4``. And maybe rephrase a bit for clarity:

    `system` needs to be a `dlti` instance or have 3, 4 or 5 arguments.

rgommers replied:
and same rephrase in the ``lti`` docstring

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -2506,3 +3130,423 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    is_ss_input = isinstance(system, StateSpace)
+    system = system._as_ss()
+
+    u = np.asarray(u)
```
``np.atleast_1d`` to avoid a confusing error for scalar input (try ``signal.dlsim((1, 2, 3), 3)``).

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -2506,3 +3130,423 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+
+    # Simulate the system
+    for i in range(0, out_samples - 1):
+        xout[i+1, :] = np.dot(system.A, xout[i, :]) +\
```
can you remove the ``\``s (replace with brackets)? it's too easy to create invalid syntax with ``\``, so we try never to use them

rgommers commented:
I'm a bit late to the party, sorry. Overall this does look good to me.

e-q commented:
Pulling the trigger, I applaud @befelix's efforts! Thanks to everyone for their input and reviews, as well.

befelix commented:
Thanks a lot @e-q and everyone who helped reviewing!
